### PR TITLE
Write release notes for Scrapy 2.0.0

### PR DESCRIPTION
**Pull request by Gallaecio**

### CODE REVIEW

File: docs/topics/asyncio.rst [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+:func:`~scrapy.utils.reactor.install_reactor`::
+
+    install_reactor('twisted.internet.asyncioreactor.AsyncioSelectorReactor')
+
+
+Coroutine support
+=================
+
+The following callables may be defined as coroutines, and hence use
+:ref:`coroutine syntax <async>` and :mod:`asyncio`, as well as
```
Comment (kmike) [inline docs/topics/asyncio.rst]: `async def` syntax should be unrelated to asyncio; I was hoping you can use it with Twisted as well, e.g. `resp = await treq.get(b"http://example.com")`.

File: scrapy/exporters.py [hunk 2]
```diff
@@ -23,7 +23,7 @@
 
 class BaseItemExporter(object):
 
-    def __init__(self, dont_fail=False, **kwargs):
+    def __init__(self, *, dont_fail=False, **kwargs):
```
Comment (kmike) [inline scrapy/exporters.py]: shouldn't it be a separate PR?

File: docs/topics/asyncio.rst (hunk 1 above)
Comment (Gallaecio) [reply docs/topics/asyncio.rst]: TIL :open_mouth:

Comment (Gallaecio) [reply docs/topics/asyncio.rst]: I’ve split out a topic about coroutines.

File: scrapy/exporters.py (hunk 2 above)
Comment (Gallaecio) [reply scrapy/exporters.py]: #4370

File: docs/topics/coroutines.rst [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+.. versionadded:: 2.0
+
+Scrapy has partial support for the :ref:`coroutine syntax <async>`.
+
+The following callables may be defined as coroutines using ``async def``, and
+hence use coroutine syntax (e.g. ``await``, ``async for``, ``async with``):
+
+-   :class:`~scrapy.http.Request` callbacks.
+
+    However, you can only use ``yield`` if you are using Python 3.6 or later.
```
Comment (kmike) [inline docs/topics/coroutines.rst]: Is this a Scrapy limitation of a Python limitation?

Comment (wRAR) [reply docs/topics/coroutines.rst]: As @Gallaecio already answered elsewhere, `yield` in an `async def` coroutine makes it an async generator and is supported only since Python 3.6.

Comment (Gallaecio) [reply docs/topics/coroutines.rst]: Maybe we should make that explicit there, or at least link to a resource that explains that?

**Comment by Gallaecio** [comment]
@whalebot-helmsman When you have time, please have a look at the [release notes section about the scheduler queue changes](https://scrapy-gallaecio.readthedocs.io/en/release-notes/news.html#changes-to-scheduler-queue-classes).

File: docs/topics/coroutines.rst (hunk 3 above)
Comment (wRAR) [reply docs/topics/coroutines.rst]: Yeah.

https://docs.python.org/3/glossary.html#term-asynchronous-generator doesn't tell the version reqs :-/

Comment (wRAR) [reply docs/topics/coroutines.rst]: We also need to mention here the limitations with the async generator callback processing from #4318

**Comment by whalebot-helmsman** [comment]
Thanks @Gallaecio for such detailed and accurate changelog.

File: docs/news.rst [hunk 4]
```diff
@@ -3,8 +3,426 @@
...
+    :class:`~scrapy.downloadermiddlewares.httpcompression.HttpCompressionMiddleware`
+    now includes spaces after commas in the value of the ``Accept-Encoding``
+    header that it sets, following web browser behavior (:issue:`4293`)
+
+*   The ``__init__`` method of download handlers (see
+    :setting:`DOWNLOAD_HANDLERS`) no longer receives a ``settings`` parameter;
+    use the ``from_settings`` or ``from_crawler`` class methods to expose such
+    a parameter to your custom download handlers (:issue:`4126`)
+
+*   Overridden settings are now logged in a different format more in line with
```
Comment (petervandenabeele) [inline docs/news.rst]: I suggest: 
`Overridden settings are now logged in a different format. This is more in line with`

File: docs/news.rst [hunk 5]
```diff
@@ -3,8 +3,426 @@
...
+    processes exceptions raised from callbacks that are generators
+    (:issue:`4260`, :issue:`4272`)
+
+*   Redirects to URLs starting with 3 slashes (``///``) are now supported
+    (:issue:`4032`, :issue:`4042`)
+
+*   :class:`~scrapy.http.Request` no longer accepts strings as ``url`` simply
+    because they have a colon (:issue:`2552`, :issue:`4094`)
+
+*   The right encoding is now used for attach names in
```
Comment (petervandenabeele) [inline docs/news.rst]: I suggest:
`The correct encoding is now used ...`

**Comment by Gallaecio** [comment]
@petervandenabeele Solved, thanks!

File: docs/news.rst [hunk 6]
```diff
@@ -3,8 +3,430 @@
...
+    :setting:`DOWNLOAD_HANDLERS`) no longer receives a ``settings`` parameter;
+    use the ``from_settings`` or ``from_crawler`` class methods to expose such
+    a parameter to your custom download handlers (:issue:`4126`)
+
+*   Overridden settings are now logged in a different format. This is more in
+    line with similar information logged at startup (:issue:`4199`)
+
+.. _Python 2 end-of-life on January 1, 2020: https://www.python.org/doc/sunset-python-2/
+
+Changes to scheduler queue classes
```
Comment (kmike) [inline docs/news.rst]: What do you think about moving all the details somewhere lower, maybe to a separate section, to highlight new features / deprecations / etc. more?

File: docs/news.rst [hunk 7]
```diff
@@ -3,8 +3,430 @@
...
+*   The new :meth:`Response.follow_all <scrapy.http.Response.follow_all>`
+    method offers the same functionality as
+    :meth:`Response.follow <scrapy.http.Response.follow>` but supports an
+    iterable of URLs as input and returns an iterable of requests
+    (:issue:`2582`, :issue:`4057`, :issue:`4286`)
+
+*   :ref:`Media pipelines <topics-media-pipeline>` now support :ref:`FTP
+    storage <media-pipeline-ftp>` (:issue:`3928`, :issue:`3961`)
+
+*   The new :attr:`Response.certificate <scrapy.http.Response.certificate>`
```
Comment (kmike) [inline docs/news.rst]: I think it can be in highlights

File: docs/news.rst (hunk 6 above)
Comment (kmike) [reply docs/news.rst]: Also, this can be in highlights, if we find a succinct way to explain the change

Comment (Gallaecio) [reply docs/news.rst]: Could you elaborate?

We could sort “highlights → new features → bug fixes → deprecation removals → deprecations → backward-incompatible changes → documentation → QA”. Do you have something else in mind?

File: docs/topics/coroutines.rst [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+.. _asynchronous generators were introduced in Python 3.6: https://www.python.org/dev/peps/pep-0525/
+
+Usage
+=====
+
+There are several use cases for coroutines in Scrapy. Code that would
+return Deferreds when written for previous Scrapy versions, such as downloader
+middlewares and signal handlers, can be rewritten to be shorter and cleaner::
+
+    class DbPipeline():
```
Comment (kmike) [inline docs/topics/coroutines.rst]:
```suggestion
    class DbPipeline:
```

File: docs/topics/coroutines.rst [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+            return item
+
+        def process_item(self, item, spider):
+            dfd = db.get_some_data(item['id'])
+            dfd.addCallback(self._update_item, item)
+            return dfd
+
+becomes::
+
+    class DbPipeline():
```
Comment (kmike) [inline docs/topics/coroutines.rst]:
```suggestion
    class DbPipeline:
```

File: docs/topics/coroutines.rst [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        async def process_item(self, item, spider):
+            item['field'] = await db.get_some_data(item['id'])
+            return item
+
+Coroutines may be used to call asynchronous libraries that use either Deferreds
+or coroutines::
```
Comment (kmike) [inline docs/topics/coroutines.rst]: I'm not sure "either Deferreds or coroutines" is precise; there are other awaitable objects which we support, e.g. asyncio Futures.

File: docs/topics/coroutines.rst [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+
+    async def parse_with_deferred(self, response):
+        additional_response = await treq.get('https://additional.url')
+        additional_data = await treq.content(additional_response)
+
+    async def parse_with_asyncio(self, response):
+        async with aiohttp.ClientSession() as session:
+            async with session.get('https://additional.url') as additional_response:
+                additional_data = await r.text()
```
Comment (kmike) [inline docs/topics/coroutines.rst]: Could you please make it a more complete example, with Spider definition? It could also make sense to show that you can output an item based on the data we're getting.

File: docs/news.rst (hunk 6 above)
Comment (kmike) [reply docs/news.rst]: I was thinking that we can say that there is a backwards incompatible change in Scheduler/queues in this section, but have a separate section at the end of the release notes, with a more detailed description. The goal is to show new features & bug fixes sooner, to get people more excited about the upgrade - backwards incompatible changs mean more work for  people, while new features & fixes mean less work. But I'm fine with merging as-is. This is nitpicking, and is super-minor.

File: docs/topics/coroutines.rst (hunk 11 above)
Comment (wRAR) [reply docs/topics/coroutines.rst]: I can add the class definition line, but a complete spider example would be too long I think. Same about the callback methods: if we want complete examples we would need to do something with both the original response and the additional one, that would make them too long, wouldn't it?

File: docs/topics/asyncio.rst [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+.. versionadded:: 2.0
+
+Scrapy has partial support :mod:`asyncio`. After you :ref:`install the asyncio
+reactor <install-asyncio>`, you may use :mod:`asyncio` and
+:mod:`asyncio`-powered libraries in any :doc:`coroutine <coroutines>`.
+
+.. warning:: :mod:`asyncio` support in Scrapy is experimental. Future Scrapy
+             versions may introduce related API and behavior changes without a
+             deprecation period or warning.
```
Comment (kmike) [inline docs/topics/asyncio.rst]: I think we can remove this warning, or reprase it; I'm not sure which API changes could we make.

File: docs/topics/coroutines.rst [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+
+        async def parse_with_deferred(self, response):
+            additional_response = await treq.get('https://additional.url')
+            additional_data = await treq.content(additional_response)
+
+        async def parse_with_asyncio(self, response):
+            async with aiohttp.ClientSession() as session:
+                async with session.get('https://additional.url') as additional_response:
+                    additional_data = await r.text()
+
```
Comment (kmike) [inline docs/topics/coroutines.rst]:
```suggestion
    class MySpider(Spider):
        # ...
        async def parse_with_deferred(self, response):
            additional_response = await treq.get('https://additional.url')
            additional_data = await treq.content(additional_response)
            # ... use response and additional_data to yield items and requests

        async def parse_with_asyncio(self, response):
            async with aiohttp.ClientSession() as session:
                async with session.get('https://additional.url') as additional_response:
                    additional_data = await r.text()
            # ... use response and additional_data to yield items and requests

```

File: docs/topics/coroutines.rst (hunk 11 above)
Comment (kmike) [reply docs/topics/coroutines.rst]: It can be a comment which says that, not a real code; added a suggestion.

File: docs/news.rst (hunk 6 above)
Comment (Gallaecio) [reply docs/news.rst]: Moved to a section at the referenciable section at the end ([rendering](https://scrapy-gallaecio.readthedocs.io/en/release-notes/news.html#scheduler-queue-changes)).

File: docs/topics/asyncio.rst (hunk 12 above)
Comment (Gallaecio) [reply docs/topics/asyncio.rst]: Changed “related API and behavior changes” to “related changes”.

**Comment by kmike** [comment]
I wonder if a test failure can be fixed by switching to a more recent pypy, hopefully without relying on third-party downloads (travis seems to provide recent pypy3 now - see https://docs.travis-ci.com/user/languages/python/#python-versions)

**Comment by wRAR** [comment]
@kmike I checked that #4366  just 9 days ago
